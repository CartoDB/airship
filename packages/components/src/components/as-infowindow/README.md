Use this component to show extra information about the data.

##Â Regular

```html
<as-infowindow>
  <h1 class="as-title">Cat</h1>
  <p>
    <span class="as-badge">mammal</span>
    <span class="as-badge as-bg--badge-pink">carnivorous</span>
  </p>
  <p class="as-body">
    The domestic cat is a small, typically furry, carnivorous mammal. They are often called house cats when kept as indoor pets or simply cats when there is no need to distinguish them from other felids and felines.
  </p>
</as-infowindow>
```

## With an image

```html
<as-infowindow src="http://on-desktop.com/wps/Animals___Cats_Black_Cat_Panther_043844_.jpg">
  <h1 class="as-title">Cat</h1>
  <p>
    <span class="as-badge">mammal</span>
    <span class="as-badge as-bg--badge-pink">carnivorous</span>
  </p>
  <p class="as-body">
    The domestic cat is a small, typically furry, carnivorous mammal. They are often called house cats when kept as indoor pets or simply cats when there is no need to distinguish them from other felids and felines.
  </p>
</as-infowindow>
```

## With only an image

```html
<as-infowindow src="http://on-desktop.com/wps/Animals___Cats_Black_Cat_Panther_043844_.jpg"></as-infowindow>
```

### CSS Variables

```
as-infowindow{
  --as--infowindow--color--background
  --as--infowindow--color--shadow
}
```

### Props

#### **src**: string
The url for the image shown at the top of the content.
