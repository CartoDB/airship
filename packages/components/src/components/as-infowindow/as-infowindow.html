<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <base href="/">
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0">
  <title>Airship</title>
  <script src="/build/airship.js"></script>
  <style>
    as-infowindow {
      margin: 8px;
    }
  </style>
</head>

<body>
  <as-infowindow src="https://thecatsite.com/styles/thecatsite/xenforo/sources/notice3.png"></as-infowindow>
  <as-infowindow src="http://funnypicture.org/wallpaper/2015/05/funny-baby-cats-11-hd-wallpaper.jpg"></as-infowindow>
  <as-infowindow src="http://www.cutestpaw.com/wp-content/uploads/2011/11/OIo.jpg"></as-infowindow>

  <as-infowindow src="http://on-desktop.com/wps/Animals___Cats_Black_Cat_Panther_043844_.jpg">
    <h1 class="as-title">Infowindow</h1>
    <p class="as-body">
      Lorem ipsum dolor sit amet consectetur adipisicing elit. Doloremque, iure sunt quia nam, totam officia labore enim corporis
      accusamus distinctio nesciunt perferendis dolore cupiditate facere non molestias expedita. Maxime, dignissimos!
    </p>
  </as-infowindow>

  <as-infowindow>
    <h1 class="as-title">Infowindow</h1>
    <p class="as-body">
      Lorem ipsum dolor sit amet consectetur adipisicing elit. Doloremque, iure sunt quia nam, totam officia labore enim corporis
      accusamus distinctio nesciunt perferendis dolore cupiditate facere non molestias expedita. Maxime, dignissimos!
    </p>
  </as-infowindow>
</body>

</html>
