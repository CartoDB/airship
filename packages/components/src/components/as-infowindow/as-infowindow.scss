@import '../../../../styles/src/core/_dev';

as-infowindow {
  --as-infowindow--color--shadow: #{rgba($color-type-01, 0.16)};
  --as-infowindow--color--background: var(--as--color--ui-01, #{$color-ui-01});

  display: inline-block;

  .as-infowindow {
    display: inline-block;
    position: relative;
    z-index: 3;
    width: 260px;
    border-radius: 4px;
    background: var(--as-infowindow--color--background);
    box-shadow: 0 4px 16px 0 var(--as-infowindow--color--shadow);

    img {
      display: block;
      width: 100%;
      border-top-left-radius: 4px;
      border-top-right-radius: 4px;
    }

    &__media {
      overflow: hidden;
      border-radius: 4px 4px 4px 0;
    }

    &__content {
      height: 100%;
      padding: 16px;
      overflow: auto;
    }

    &__hook {
      position: absolute;
      z-index: 10;
      bottom: 1px;
      left: 0;
      background: var(--as-infowindow--color--background);

      &::before {
        content: '';
        position: absolute;
        z-index: 3;
        top: 0;
        left: 0;
        width: 0;
        height: 0;
        border-top: 12px solid var(--as-infowindow--color--background);
        border-right: 18px solid transparent;
        border-radius: 0 0 0 4px;
      }

      &::after {
        content: '';
        position: absolute;
        z-index: 2;
        top: 3px;
        left: 0;
        width: 0;
        height: 0;
        border-top: 12px solid var(--as-infowindow--color--shadow);
        border-right: 18px solid transparent;
        filter: blur(2px);
      }
    }

    &__imageHook {
      position: absolute;
      z-index: 10;
      bottom: 0;

      &::after {
        content: '';
        position: absolute;
        z-index: 1;
        top: 3px;
        left: 0;
        width: 0;
        height: 0;
        border-top: 12px solid var(--as-infowindow--color--shadow);
        border-right: 18px solid transparent;
        filter: blur(2px);
      }
    }

    &__imageHook--inner {
      position: absolute;
      z-index: 3;
      top: -1px;
      width: 18px;
      height: 12px;
      overflow: hidden;
      clip-path: polygon(0 0, 0% 100%, 100% 0);

      & > img {
        width: 260px;
      }
    }
  }
}
