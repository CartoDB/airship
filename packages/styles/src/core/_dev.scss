@import 'tools/_mixins';
@import 'variables/_variables';
