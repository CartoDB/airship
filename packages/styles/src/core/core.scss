@import 'variables/_variables';

* {
  box-sizing: border-box;
  font-family: sans-serif;
}

:root {
  --primary-color: #{$primary-color};
}
