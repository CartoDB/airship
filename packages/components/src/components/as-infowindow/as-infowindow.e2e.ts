import { newSpecPage } from '@stencil/core/testing';
import { Infowindow } from './as-infowindow';

describe('as-infowindow', () => {
  describe('Rendering', () => {
    it('should render content', async () => {
      const page = await newSpecPage({
        components: [Infowindow],
        html: `
        <as-infowindow>Hello there!</as-infowindow>
      `});

      const element = page.root;
      expect(element.outerHTML).toMatchSnapshot();
    });

    it('should render content and image', async () => {
      const page = await newSpecPage({
        components: [Infowindow],
        html: `
        <as-infowindow src="general_kenobi.jpg">Hello there!</as-infowindow>
      `});

      const element = page.root;
      expect(element.outerHTML).toMatchSnapshot();
    });

    it('should render with only an image', async () => {
      const page = await newSpecPage({
        components: [Infowindow],
        html: `
        <as-infowindow src="general_kenobi.jpg"></as-infowindow>
      `});

      const element = page.root;
      expect(element.outerHTML).toMatchSnapshot();
    });
  });
});
