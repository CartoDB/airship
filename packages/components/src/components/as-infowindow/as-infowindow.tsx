import { Component, Element, h, Prop, } from '@stencil/core';

@Component({
  shadow: false,
  styleUrl: './as-infowindow.scss',
  tag: 'as-infowindow',
})

export class Infowindow {

  @Prop() public src: string;
  @Element() private element: HTMLElement;

  public render() {
    return this.element.childElementCount === 0 && this.src
      ? this._renderImageInfoWindow()
      : this._renderStandarInfowindow();
  }

  public componentDidLoad() {
    if (this.src) {
      this._setupHook();
    }
  }

  private _renderStandarInfowindow() {
    return (
      <div class='as-infowindow'>
        {this.src && <img src={this.src} />}
        <div class='as-infowindow__content'>
          <slot />
        </div>
        <div class='as-infowindow__hook'></div>
      </div>
    );
  }

  private _renderImageInfoWindow() {
    return (
      <div class='as-infowindow'>
        <div class='as-infowindow__media'>
          <img src={this.src} />
        </div>
        <div class='as-infowindow__imageHook'>
          <div class='as-infowindow__imageHook--inner'>
            <img src={this.src} />
          </div>
        </div>
      </div>
    );
  }

  private _setupHook() {
    const imageElement = this.element.querySelector('.as-infowindow__media img') as HTMLImageElement;

    if (!imageElement) {
      return;
    }

    imageElement.onload = () => {
      const imageHeight = imageElement.offsetHeight;

      const media = this.element.querySelector('.as-infowindow__media') as HTMLDivElement;
      const hook = this.element.querySelector('.as-infowindow__imageHook img') as HTMLDivElement;

      const offset = imageHeight - 18;

      media.style.height = `${offset}px`;
      hook.style.marginTop = `-${offset}px`;
    };
  }
}
