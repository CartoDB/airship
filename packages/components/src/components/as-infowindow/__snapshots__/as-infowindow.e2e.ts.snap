// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`as-infowindow Rendering should render content 1`] = `"<as-infowindow><!----> <div class=\\"as-infowindow\\"><div class=\\"as-infowindow__content\\"> Hello there!</div><div class=\\"as-infowindow__hook\\"></div></div></as-infowindow>"`;

exports[`as-infowindow Rendering should render content and image 1`] = `"<as-infowindow src=\\"general_kenobi.jpg\\"><!---->Hello there!<div class=\\"as-infowindow\\"><div class=\\"as-infowindow__media\\"><img src=\\"general_kenobi.jpg\\"></div><div class=\\"as-infowindow__imageHook\\"><div class=\\"as-infowindow__imageHook--inner\\"><img src=\\"general_kenobi.jpg\\"></div></div></div></as-infowindow>"`;

exports[`as-infowindow Rendering should render with only an image 1`] = `"<as-infowindow src=\\"general_kenobi.jpg\\"><!----><div class=\\"as-infowindow\\"><div class=\\"as-infowindow__media\\"><img src=\\"general_kenobi.jpg\\"></div><div class=\\"as-infowindow__imageHook\\"><div class=\\"as-infowindow__imageHook--inner\\"><img src=\\"general_kenobi.jpg\\"></div></div></div></as-infowindow>"`;
